csi.proto

```diff
@@ -1634,3 +1679,190 @@ message NodeExpandVolumeResponse {
   int64 capacity_bytes = 1;
 }
+message GroupControllerGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating and deleting a volume group snapshot.
+      CREATE_DELETE_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+
+      // Indicates whether the group controller plugin supports
+      // creating a volume from an individual volume snapshot if the
+      // volume snapshot is part of a VolumeGroupSnapshot.
+      // Use cases: selective restore, advanced recovery, etc.
+      INDIVIDUAL_SNAPSHOT_RESTORE = 2 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a volume group snapshot.
+      GET_VOLUME_GROUP_SNAPSHOT = 3 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a list of volume group snapshots.
+      LIST_VOLUME_GROUP_SNAPSHOTS = 4 [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // suggested name for a group snapshot (required for idempotent)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // volume ids of the volumes to be snapshotted together.
+  // This field is REQUIRED.
+  repeated string volume_ids = 2;
+
+  // Secrets required by plugin to complete volume group snapshot
+  // creation request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 3 [(.csi.v1.csi_secret) = true];
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  map<string, string> parameters = 4;
+}
+
+message CreateVolumeGroupSnapshotResponse {
+  option (alpha_message) = true;
+
+  // Contains all attributes of the newly created group snapshot.
+  // This field is REQUIRED.
+  VolumeGroupSnapshot group_snapshot = 1;
+}
+
+message VolumeGroupSnapshot {
+  option (alpha_message) = true;
+
+  // The identifier for this group snapshot, generated by the plugin.
+  // This field is REQUIRED.
+  string group_snapshot_id = 1;
+
+  // A list of snapshots created. Snapshot is the same
+  // definition as Snapshot definition used in CSI.
+  // This field is OPTIONAL.
+  repeated Snapshot snapshots = 2;
```

Comment:
According to feedback from @humblec on the VolumeGroup KEP, CephFS does not return individual snapshot ids. It only returns a group snapshot id.  For CephFS, a volume group is a filesystem.  See more discussions here:

https://github.com/kubernetes/enhancements/pull/1551#issuecomment-878805651
https://github.com/kubernetes/enhancements/pull/1551#discussion_r687683605

So CO won't be able to store individual snapshot records when taking a GroupSnapshot for a storage system such as CephFS. At restore time, since we no longer have a VolumeGroup in the latest proposal, user won't be able to create a volume from an individual snapshot since it is not available. User will have to restore the group (filesystem in this case) from the group snapshot out of band. PVCs will be statically provisioned from that filesystem. I have asked @humblec for clarifications on how to restore if we don't have a VolumeGroup and cannot create a VolumeGroup from a VolumeGroupSnapshot.

@humblec Please chime in.  Thanks.

Reply:
I'm not really okay with supporting workflows where only half the workflow can be done through CSI and the other half must be done out of band. We should either ensure that the workflow is entirely supported, or we should leave it out and force everything to be done out of band.

Ceph may be one of those unusual examples I was alluding to where group snapshots can be implemented, but only with serious limitations. I recall similar situations from my previous job where the only way we could have supported group snapshots was if the group concept was extremely constrained.

We should be careful about supporting such limited implementations if doing so detracts from the user experience. I would rather keep a high bar for user experience even if that means some storage systems can't implement the feature, as long as doing so doesn't exclude a significant plurality of implementations.

It's always possible that storage vendors can make their products better to meet the bar we set. If we create a bad user experience, there is no way to recover from that though.

Since I don't really understand Ceph though I don't want to assume it has any specific limitations. Let's consider if there's a way we change the design such that Ceph can fully implement it without significantly reducing how useful this feature is in general.

Reply:
I have two proposals:
1. Make snapshot_ids a required field. Maybe CephFS CSI driver can return <group_snapshot_id>+<volume_id> if it can't return an individual snapshot_id? At restore time, user will try to create a volume from a snapshot. CephFS CSI driver will first create the filesystem (skip this step if filesystem is already created), then create a volume based on <group_snapshot_id>+<volume_id>.
2. Add VolumeGroup back. In this case, snapshot_ids are still optional. If not provided, individual snapshots are not created. At restore time, a VolumeGroup will be created from a VolumeGroupSnapshot and individual volumes will be created and added to a group.

@humblec Please verify if both approaches work for CephFS.

Reply:
For now, I'm going with option 1 and made "snapshots" a required field in VolumeGroupSnapshot message.

Reply:
First of all, thanks @xing-yang  for accommodating the suggestions/reviews from different angles since the very first version of the KEP to this version :), its not an easy task!! 

Coming back to the new (recent) version of this spec ie `without volumegroup` and the APIs it currently expose, I feel `option 1` is better or viable option.  That said, in the return/response of create group snapshot request, for the snapshot IDs field, the SP driver has the choice of either using the `volumeIds` from the request itself or even it can append/hash the internal snap details ( id, name, location ..etc) / any other information with/without  the volume IDs or with/without group snapshot request name  it received in the request. That model looks really flexible to me. Later,  at any point in time, when the restore is requested on a snapshotID of a  group snapshot , the SP driver is  in a position to restore the volume to new one either directly or by deriving its details from the (encoded) snapshot ID.  In that model, ideally any SP driver should be able to support this workflow.  more or less, I am talking about SP driver in general and not specifically Ceph. imo,  this looks like a workable solution for Ceph as well, however I will continue the discussions with Ceph core team about this model and provide the feedback. 
 
In absence of volumegroup or clubbing volume group APIs with group snapshot, this proposal looks bit more simple in terms of , the combination of the operations each component ( group snapshot controller / CO/ CSI driver ) has to handle. It seems that, this proposal has turned out to be a hook for the backend storage system to act and get point in time copy of a group of volumes which are bound to a workload/specific use case. iow, most of the work has been delegated to storage subsystem which is good and exactly what this KEP has to address :+1: